ui-library-filesharing-ui-components/app/src/ChatComponents.tsx

```diff
@@ -0,0 +1,137 @@
+import {
+  usePropsFor,
+  MessageThread,
+  SendBox,
+  ActiveFileUpload,
+  FileMetadata,
+} from "@azure/communication-react";
+import React from "react";
+import axios from "axios";
+import Form from "form-data";
+import { v4 } from "uuid";
+
+function ChatComponents(): JSX.Element {
+  const allActiveFileUploads = React.useRef<ActiveFileUpload[] | []>([]);
+  const completedFileUploads = React.useRef<FileMetadata[] | []>([]);
+  const [files, setFiles] = React.useState<File[] | []>();
+  const [activeFileUploads, setActiveFileUploads] = React.useState<
+    ActiveFileUpload[] | []
+  >([]);
+
+  const updateFileUploadProgress = (
+    fileId: string,
+    progress: number,
+    complete: boolean = false
+  ) => {
+    const updatedData = allActiveFileUploads.current.map((active) => {
+      if (active.id === fileId) {
+        return {
+          ...active,
+          progress,
+          uploadComplete: complete,
+        };
+      }
+      return active;
+    });
+    allActiveFileUploads.current = updatedData;
+    setActiveFileUploads(updatedData);
+  };
+
+  const uploadFile = async (file: File): Promise<void> => {
+    const extension = file.name.split(".").pop() || "";
```

Comment:
Once again adding a couple helpers can make this easier to follow

```
const updateFileProgress = (p) => updateFileUploadProgress(file.name, p.loaded/p.total);
triggerUpload(file, updateFileProgress)
    .then((res) => {
        updateFileUploadProgress(file.name, 1, true);
        completedFileUploads.current = appendCompletedFileUpload(completedFileUploads.current, file, res);
    }). catch(err) => {
        console.log(err);
    })
```

Here `triggerUpload` and `appendCompletedFileUpload` are file-level private functions.

Reply:
This seems like overkill to me. I think too much abstraction is a bad abstraction.

Reply:
Let's get a third opinion (that's what we have multiple reviewers for ;) )

It isn't about abstraction. The question I ask is: Do I need to look at how axios and multipart work when trying to understand the state keeping in this component. Conversely, would be beneficial to read the axios/multipart code without the distraction of React component statekeeping?

By extracting `triggerUpload`, I think you can separate out two very different concerns when reading the code.

Reply:
I understand your POV. I have a slightly different personal preference.
I like seeing everything in one place when I look at samples instead of me having to jump around function declarations to understand what's doing what...

Let's see what @alkwa-msft thinks